### PR TITLE
MYNEWT-715 Extended Advertising - Advertiser support for controller

### PR DESCRIPTION
**Pull request by sjanc**
This adds support for controller for Extended Advertiser. Included:
 - all HCI commands and events for setting advertising instances
 - support for all types of AUX advertising PDUs
 - advertising using Coded and 2M PHY
 - Android extensions are removed in favor of new commands
 
Limitations (will be address in followup PRs):
 - Advertising Data and Scan_Rsp are still limited to 31 bytes
 - no chaining support (only 1 AUX_ADV_IND per advertising event)
 - advertising duration parameter is ignored (only events count is implemented)

### CODE REVIEW

**Comment by wes3** [comment]
Changes looks good to me.

**Comment by sjanc** [comment]
I've updated PR branch.  Fixed nitpick pointed by WIll and fixed sending scan req notification to host.

File: net/nimble/controller/src/ble_ll_adv.c [hunk 1]
```diff
@@ -547,17 +907,165 @@ ble_ll_adv_set_sched(struct ble_ll_adv_sm *advsm)
...
+    assert(rc == 0);
+
+    /* Set transmit start time. */
+#if MYNEWT_VAL(OS_CPUTIME_FREQ) == 32768
+    txstart = sch->start_time + g_ble_ll_sched_offset_ticks;
+    rc = ble_phy_tx_set_start_time(txstart, sch->remainder);
+#else
+    txstart = sch->start_time +
+        os_cputime_usecs_to_ticks(XCVR_TX_SCHED_DELAY_USECS);
+    rc = ble_phy_tx_set_start_time(txstart);
```
Comment (rymanluk) [inline net/nimble/controller/src/ble_ll_adv.c]: this can be removed

File: net/nimble/controller/src/ble_ll_adv.c [hunk 2]
```diff
@@ -547,17 +907,165 @@ ble_ll_adv_set_sched(struct ble_ll_adv_sm *advsm)
...
+        max_usecs += BLE_LL_SCHED_ADV_MAX_USECS;
+    }
+
+    /*
+     * XXX: For now, just schedule some additional time so we insure we have
+     * enough time to do everything we want.
+     */
+    max_usecs += XCVR_PROC_DELAY_USECS;
+
+#if MYNEWT_VAL(OS_CPUTIME_FREQ) == 32768
```
Comment (rymanluk) [inline net/nimble/controller/src/ble_ll_adv.c]: ditto

File: net/nimble/controller/src/ble_ll_adv.c [hunk 3]
```diff
@@ -779,6 +1262,57 @@ ble_ll_adv_sm_stop(struct ble_ll_adv_sm *advsm)
     }
 }
 
+static void
+ble_ll_adv_scheduled(struct ble_ll_adv_sm *advsm, uint32_t sch_start)
+{
+#if MYNEWT_VAL(OS_CPUTIME_FREQ) == 32768
+    /* The event start time is when we start transmission of the adv PDU */
+    advsm->adv_event_start_time = sch_start + g_ble_ll_sched_offset_ticks;
+#else
```
Comment (rymanluk) [inline net/nimble/controller/src/ble_ll_adv.c]: no needed

File: net/nimble/controller/src/ble_ll_adv.c [hunk 4]
```diff
@@ -894,228 +1418,676 @@ ble_ll_adv_sm_start(struct ble_ll_adv_sm *advsm)
...
+
+    advsm = &g_ble_ll_adv_sm[instance];
+    datalen = cmd[0];
+
+    /* check if type of advertising support scan rsp */
+    if (!(advsm->props & BLE_HCI_LE_SET_EXT_ADV_PROP_SCANNABLE)) {
+        return BLE_ERR_INV_HCI_CMD_PARMS;
+    }
+
+    if (!(advsm->props & BLE_HCI_LE_SET_EXT_ADV_PROP_SCANNABLE)) {
```
Comment (rymanluk) [inline net/nimble/controller/src/ble_ll_adv.c]: Looks like redundant double check

File: net/nimble/controller/src/ble_ll_adv.c [hunk 5]
```diff
@@ -894,228 +1418,676 @@ ble_ll_adv_sm_start(struct ble_ll_adv_sm *advsm)
...
+
+        break;
+#if MYNEWT_VAL(BLE_LL_CFG_FEAT_LL_EXT_ADV)
+    case BLE_HCI_LE_SET_EXT_SCAN_RSP_DATA_OPER_UNCHANGED:
+        if (advsm->props & BLE_HCI_LE_SET_EXT_ADV_PROP_LEGACY) {
+            return BLE_ERR_INV_HCI_CMD_PARMS;
+        }
+
+        if (!advsm->adv_enabled || !advsm->adv_len || datalen) {
+            return BLE_ERR_INV_HCI_CMD_PARMS;
```
Comment (rymanluk) [inline net/nimble/controller/src/ble_ll_adv.c]: just asking. below on !adv_enabled we send back command_disallowed and here invalid parameters. is that ok?

Comment (sjanc) [inline net/nimble/controller/src/ble_ll_adv.c]: yeah,  some operations are not allowed if advertising is enabled,   but I just noticed that operation 0x04 is valid only for setting advertising data, not scan_rsp. I'lld fix that.

File: net/nimble/include/nimble/hci_common.h [hunk 6]
```diff
@@ -458,19 +458,49 @@ extern "C" {
...
+#define BLE_HCI_LE_SET_EXT_ADV_PROP_CONNECTABLE     (0x0001)
+#define BLE_HCI_LE_SET_EXT_ADV_PROP_SCANNABLE       (0x0002)
+#define BLE_HCI_LE_SET_EXT_ADV_PROP_DIRECTED        (0x0004)
+#define BLE_HCI_LE_SET_EXT_ADV_PROP_HD_DIRECTED     (0x0008)
+#define BLE_HCI_LE_SET_EXT_ADV_PROP_LEGACY          (0x0010)
+#define BLE_HCI_LE_SET_EXT_ADV_PROP_ANON_ADV        (0x0020)
+#define BLE_HCI_LE_SET_EXT_ADV_PROP_INC_TX_PWR      (0x0040)
+#define BLE_HCI_LE_SET_EXT_ADV_PROP_MASK            (0x7F)
+
+#define BLE_HCI_LE_SET_EXT_ADV_PROP_LEGACY_IND      (0x0013)
```
Comment (sjanc) [inline net/nimble/include/nimble/hci_common.h]: only first 5 props match what is set also in extended advertising event, but those are for separate command and (there is also anon adv and tx power which are not present in event), also mask size is different,

I'd leave it as is

**Comment by sjanc** [comment]
I've updated branch with comments addresses (except flags define which I left as is)

**Comment by rymanluk** [comment]
Thanks, looks good.

**Pull request merged**